src/main/cpp/src/cast_string_to_float.cu

```diff
@@ -0,0 +1,716 @@
+    _c      = _warp_lane < _blen ? _chars[_row_start + _warp_lane] : 0;
+
+    if (_incoming_null_mask != nullptr && !bit_is_set(_incoming_null_mask, _row)) {
+      _valid = false;
+      compute_validity(_valid, _except);
+      return;
+    }
+
+    remove_leading_whitespace();
+
+    // check for + or -
+    int sign = check_for_sign();
+
+    // check for leading nan
+    if (check_for_nan()) {
+      _out[_row] = NAN;
+      compute_validity(_valid, _except);
+      return;
+    }
+
+    // check for inf / infinity
+    if (check_for_inf()) {
+      if (_warp_lane == 0) {
+        _out[_row] =
+          sign > 0 ? std::numeric_limits<T>::infinity() : -std::numeric_limits<T>::infinity();
+      }
+      compute_validity(_valid, _except);
+      return;
+    }
+
+    // parse the remainder as floating point.
+    auto const [digits, exp_base] = parse_digits();
+    if (!_valid) {
+      compute_validity(_valid, _except);
+      return;
+    }
+
+    // 0 / -0.
+    if (digits == 0) {
+      remove_leading_whitespace();
+      if (_bpos < _blen) {
+        _valid  = false;
+        _except = true;
+      }
+
+      if (_warp_lane == 0) { _out[_row] = sign * static_cast<double>(0); }
+      compute_validity(_valid, _except);
+      return;
+    }
+
+    // parse any manual exponent
+    auto const manual_exp = parse_manual_exp();
+    if (!_valid) {
+      compute_validity(_valid, _except);
+      return;
+    }
+
+    check_trailing_bytes();
+    if (!_valid) {
+      compute_validity(_valid, _except);
+      return;
+    }
+
+    // construct the final float value
+    if (_warp_lane == 0) {
+      // base value
+      double digitsf = sign * static_cast<double>(digits);
+
+      // exponent
+      int exp_ten = exp_base + manual_exp;
+
+      // final value
+      if (exp_ten > std::numeric_limits<double>::max_exponent10) {
+        _out[_row] = sign > 0 ? std::numeric_limits<double>::infinity()
+                              : -std::numeric_limits<double>::infinity();
+      } else {
+        // make sure we don't produce a subnormal number.
+        // - a normal number is one where the leading digit of the floating point rep is not zero.
+        //      eg:   0.0123  represented as  1.23e-2
+        //
+        // - a denormalized number is one where the leading digit of the floating point rep is zero.
+        //      eg:   0.0123 represented as   0.123e-1
+        //
+        // - a subnormal number is a denormalized number where if you tried to normalize it, the
+        // exponent
+        //   required would be smaller then the smallest representable exponent.
+        //
+        // https://en.wikipedia.org/wiki/Denormal_number
+        //
+
+        double const exponent = exp10(static_cast<double>(std::abs(exp_ten)));
+        double const result   = exp_ten < 0 ? digitsf / exponent : digitsf * exponent;
+
+        _out[_row] = result;
```

Comment:
This may do a narrow cast and issue a compiler warning. So we need explicit `static_cast<T>` here.